src/helpers/counterBuilder.js

```diff
@@ -0,0 +1,229 @@
+    } else if (typeof this.props.maxDigits === 'object') {
+      maxDigits = this.props.maxDigits
+    }
+    return [minDigits, maxDigits]
+  }
+
+  /**
+   * Calculates minimum number of digits for current radix and given period.
+   * @param {string} period - period to calculate minDigits for
+   * @return {number}
+   */
+  getMinSegmentSize (period) {
+    const maxValue = this.getPeriodMaxValue(period)
+    return (maxValue).toString(this.props.radix).length
+  }
+
+  /**
+   * Normalizes minDigits and maxDigits and sets default values where user did not specify them.
+   * @param {number|object} minDigits
+   * @param {number|object} maxDigits
+   * @return {object}
+   */
+  normalizeDigitLimits (minDigits, maxDigits) {
+    var normalizedMinDigits = {}
+    var normalizedMaxDigits = {}
+
+    for (let period of this.state.periods) {
+      const minSegmentSize = this.getMinSegmentSize(period);
+      [normalizedMinDigits[period], normalizedMaxDigits[period]] = this.normalizeDigitLimitsForPeriod(
+        minDigits[period], maxDigits[period], minSegmentSize
+      )
+    }
+    return {
+      minDigits: normalizedMinDigits,
+      maxDigits: normalizedMaxDigits
+    }
+  }
+
+  normalizeDigitLimitsForPeriod (min, max, minSegmentSize) {
+    const minProvided = min !== undefined
+    const maxProvided = max !== undefined
+    function throwError () {
+      throw new Error(`conflict: minDigits (${min}) > maxDigits (${max})`)
+    }
+
+    if (minProvided && max) {
+      return (min > max) ? throwError() : [min, max]
+    } else if (minProvided && !maxProvided) {
+      return [min, (min > minSegmentSize) ? min : minSegmentSize]
+    } else if (!minProvided && max) {
+      return [(minSegmentSize > max) ? max : minSegmentSize, max]
+    } else {
+      return [minSegmentSize, maxProvided ? max : minSegmentSize]
+    }
+  }
+
+  /**
+   * Calculates maximum value for given period.
+   * @param {string} period
+   * @return {number}
+   */
+  getPeriodMaxValue (period) {
+    switch (period) {
+      case 'seconds':
+      case 'minutes':
+        return 59
+      case 'hours':
+        return 23
+      default:
+        if (period === this.props.minPeriod) {
+          // max number possible for this period during the countdown
+          return NumberCalculator.getPeriodNumberAt(
+            period, this.props.maxPeriod, this.state.initialTimeDiff
+          )
+        } else {
+          return this.getPeriodMaxValue(this.props.minPeriod)
+        }
+    }
+  }
+
+  /**
+   * Calculates initial numbers for all periods.
+   */
+  buildInitialNumbers () {
+    const timestamp = (this.props.direction === 'down') ? this.state.initialTimeDiff : 0
+    this.state = {
+      ...this.state,
+      numbers: NumberCalculator.calculateNumbers(this.state.periods, timestamp)
+    }
+  }
+
+  /**
+   * Validates counter's props.
+   * @param {Object} props - props to validate
+   */
+  static validateProps (props) {
```

Comment:
In ruby, there is "Static methods should be declared first" rule. I find it good